File: Data Structure/Heap/LET_no1675_Minimize Deviation in Array.py
import heapq

class Solution:
    def minimumDeviation(self, nums: List[int]) -> int:
        pq = []
        for num in nums:
            heapq.heappush(pq, (num // (num & -num), num))
        
        res = sys.maxsize
        maxVal = max(num2 for num2, num in pq)
        
        while len(pq) == len(nums):
            num2, num = heapq.heappop(pq)
            res = min(res, maxVal - num2)
            if num2 % 2 != 0 or num2 < num:
                num2 *= 2
                maxVal = max(maxVal, num2)
                heapq.heappush(pq, (num2, num))
        
        return res

File: Data Structure/Heap/LET_no1675_Minimize Deviation in Array.cpp
class Solution {
public:
    int minimumDeviation(vector<int>& nums) {
        int res = INT_MAX, min_val = INT_MAX, cur;
        priority_queue<int> pq;
        for (int num : nums) {
            if (num % 2)
                num *= 2;
            pq.push(num);
            min_val = min(min_val, num);
        }
        while (pq.top() % 2 == 0) {
            cur = pq.top();
            res = min(res, cur - min_val);
            min_val = min(min_val, cur / 2);
            pq.push(cur / 2);
            pq.pop();
        }

        return min(res, pq.top() - min_val);        
    }
};